File: day:31/buySellOnce.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    // base case
    if(prices.length === 0) return 0;
    
    let smallestSoFar = prices[0];
    let maxProfitSoFar = -Infinity;
    
    for(let price of prices) {
        smallestSoFar = Math.min(smallestSoFar, price);
        maxProfitSoFar = Math.max(maxProfitSoFar, (price - smallestSoFar));
    }
    
    return maxProfitSoFar;
    
};

File: day:31/buySellOnce.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        #base case
        if(len(prices) == 0): return 0;
    
        smallestSoFar = prices[0];
        maxProfitSoFar = -sys.maxsize;
    
        for price in prices:
            smallestSoFar = min(smallestSoFar, price);
            maxProfitSoFar = max(maxProfitSoFar, (price - smallestSoFar));
    
        return maxProfitSoFar;
